text/0000-optimize-by-default.md

```diff
@@ -0,0 +1,185 @@
+[a user accidentally compiling C (and Haskell) code without optimizations][ex5]
+and then comparing with Rust.
+
+**Debug-by-default is a poor design.** It's a footgun of massive proportions.
+Rust is already breaking with programming convention in a myriad ways to
+eliminate footguns (which is excellent); it should break with convention here as
+well.
+
+**Let's not leave footguns in place because of obsolete conventions created by
+backwards compatibility decisions made 40 years ago that aren't relevant to
+Rust.**
+
+## Why should we do this before 1.0?
+
+Because doing it after 1.0 breaks backwards compatibility on a command-line
+level. People will expect `rustc`/`cargo` to behave a certain way and will encode
+those assumptions into their bash command history/build files/scripts/whatever.
+
+# Detailed design
+
+Change `rustc` and `cargo` to produce optimized binaries by default.
+
+`cargo` currently has a `--release` flag; it should also get a `--debug` flag to
+trigger a debug build. By default, running `cargo build` is the same as running
+`cargo build --release`.
+
+Similarly, `rustc` should by default run with the same flags that it would
+receive from `cargo` if it were passed `--release`. It might make sense to also
+add support for `--release` & `--debug` flags to `rustc`, so that with
+`--debug` it becomes easy to instruct the compiler to not include optimizations
+and to include debug info.
+
+Debate is welcome on the specifics as long as "unadorned" `rustc` and `cargo
+build` calls produce optimized binaries.
+
+# Drawbacks
+
+## Slower builds by default
+
+Builds will only be slower _by default_; in other words, anyone calling `rustc
+foo.rs` today can trivially call `rustc -C optlevel=0 foo.rs` or `rustc --debug
+foo.rs` if the extra `rustc` flags are implemented.
+
+Note the failure state of debug-by-default: if you wanted an optimized build
+instead, your production binary is now incredibly slow. As someone who has
+actually pushed unoptimized binaries to production by accident, I can
+confidently say _this costs a lot of money._
+
+For opt-by-default, if you actually wanted a debug build the failure state is
+"you waited longer for your build to finish than you should have," which while
+annoying, has _far_ less potential for damage.
+
+## Breaks convention with compilers for other languages
+
+While this is a real concern, it's the result of those compilers continuing to
+carry a legacy burden. C compilers need to be backwards compatible with the
+state-of-the-art from many decades ago, and C++ compilers need to be compatible
+with C compilers. In other words, they don't have a choice.
+
+Rust has been breaking plenty of language conventions for the sake of both
+memory-safety and safety in general; it should continue on this commendable path
+without being hamstrung by decisions made in 1972.
+
+# Alternatives
+
+## Only have `cargo` echo a message stating `debug` or `release` build mode
```

Comment:
This is not something that I would consider "unsafe". "Unexpected by some" would be my preferred phrasing. Also, just because people come from other languages doesn't mean that rust needs to cater to their expectations.

I, for one, definitely expect to get a debug build with the "minimal-typing way to build my stuff" command. If it does otherwise, I'll be a little bit surprised, but as long as the build system properly communicates what it's doing, that's fine. Right now, we don't communicate the build type at all. If that's fixed, all should be right with the world.

Points that I do agree with:

Put debug binaries in `target/debug/` instead of `target/`. Clear the "other" directory on build unless both debug and release flags are passed.
Embed the word "unoptimized" or "unoptimized" in the build output.